frame/bridge/ethereum/relay/src/lib.rs

```diff
@@ -1,67 +1,261 @@
+			}
+		}
 	}
 }
 
-impl<T: Trait<I>, I: Instance> Relayable for Module<T, I> {
+impl<T: Trait> Module<T> {
+	/// validate block with the hash, difficulty of confirmed headers
+	fn verify_block_with_confrim_blocks(header: &EthHeader) -> bool {
+		let eth_partial = EthashPartial::production();
+		let confirm_blocks_in_cycle = ConfirmBlocksInCycle::get();
+		if ConfirmedHeadersDoubleMap::contains_key(
+			(header.number.saturating_sub(1)) / confirm_blocks_in_cycle,
+			header.number.saturating_sub(1),
+		) {
+			let previous_header = ConfirmedHeadersDoubleMap::get(
+				(header.number.saturating_sub(1)) / confirm_blocks_in_cycle,
+				header.number.saturating_sub(1),
+			);
+			if header.parent_hash != previous_header.hash.unwrap_or_default()
+				|| *header.difficulty()
+					!= eth_partial.calculate_difficulty(header, &previous_header)
+			{
+				return false;
+			}
+		}
+
+		if ConfirmedHeadersDoubleMap::contains_key(
+			(header.number.saturating_add(1)) / confirm_blocks_in_cycle,
+			header.number.saturating_add(1),
+		) {
+			let subsequent_header = ConfirmedHeadersDoubleMap::get(
+				(header.number.saturating_add(1)) / confirm_blocks_in_cycle,
+				header.number.saturating_add(1),
+			);
+			if header.hash.unwrap_or_default() != subsequent_header.parent_hash
+				|| *subsequent_header.difficulty()
+					!= eth_partial.calculate_difficulty(&subsequent_header, header)
+			{
+				return false;
+			}
+		}
+
+		true
+	}
+
+	fn verify_block_seal(header: &EthHeader, ethash_proof: &[DoubleNodeWithMerkleProof]) -> bool {
+		if header.hash() != header.re_compute_hash() {
+			return false;
+		}
+
+		let eth_partial = EthashPartial::production();
+
+		if eth_partial.verify_block_basic(header).is_err() {
+			return false;
+		}
+
+		if eth_partial.verify_block_basic(header).is_err() {
+			return false;
+		}
+
+		// NOTE:
+		// The `ethereum-linear-relay` is ready to drop, the merkle root data will migrate to
+		// `ethereum-relay`
+		let merkle_root = <Module<T>>::dag_merkle_root((header.number as usize / 30000) as u64);
+
+		if eth_partial
+			.verify_seal_with_proof(&header, &ethash_proof, &merkle_root)
+			.is_err()
+		{
+			return false;
+		};
+
+		return true;
+	}
+
+	/// Verify the `leaf_hash` is included in the mmr tree, which root of tree is `mmr_root`.
+	/// The `mmr_proof` is a serial of hash in the path from `leaf_hash` to `mmr_root`
+	/// The length of `mmr_proof` is linear related with the height of mmr tree.
+	///
+	///                       mmr_root
+	///                      /
+	///                   ...
+	///                 /    \
+	///                -      nth hash of mmr_proof
+	///              /   \
+	///             -     3rd hash of mmr_proof
+	///           /   \
+	///          -     2nd hash of mmr_proof
+	///        /   \
+	/// leaf_hash   1st hash of mmr_proof
+	///
+	/// mmr_proof = [ 1st_hash, 2nd_hash, 3rd_hash ... nth_hash ]
+	///
+	fn verify_mmr(
```

Comment:
Need to check the spec of the MMRProof here with the proof generated by nervos lib.

Reply:
Implementation need to use nervos lib's verify:

https://github.com/nervosnetwork/merkle-mountain-range/blob/ea232474496bdf072d7de33297fd54f1ef8df599/src/mmr.rs#L271

Reply:
The two functions are used to generate MerkleProof and verify the proof within nervos lib.
```pub fn new(mmr_size: u64, proof: Vec<T>) -> MerkleProof ```
```pub fn verify(&self, root: T, leaves: Vec<(u64, T)>) -> Result<bool> ```

Following parameters are needed from relayer.
1. proof
2. all leaves with positions
3. mmr root

`mmr_size` is assumed be calculated as mentioned in https://github.com/darwinia-network/darwinia-common/issues/177,
if not  the mmr_size also need to be provided by relayer.